experimental/serde_unaligned/src/uvec/mod.rs

```diff
@@ -0,0 +1,321 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+#[cfg(feature = "serde")]
+mod serde;
+
+use crate::byte_slice::*;
+use std::borrow::Cow;
+
+/// The inner representation of a `UVec`. A `UVec` cannot be constructed directly from a
+/// `UVecInner` because data may need validation; use the `From` impls on `UVec` instead.
+// TODO: Consider a custom Debug impl
+#[derive(Debug)]
+pub enum UVecInner<'a, T> {
+    Owned(Vec<T>),
+    Aligned(&'a [T]),
+    /// Buffer with little-endian data
+    UnalignedLE(&'a [u8]),
```

Comment:
I wouldn't do it this way. I'd instead have a `AsUnaligned` trait with a `T::Unaligned` type, which for `u32` would be `UU32`, for `TinyStr` would be `UTinyStr`, etc.

Reply:
```rust
/// invariants: the conversions are simple transmutes
/// `T::Unaligned` has `u8` alignment
unsafe trait AsUnaligned {
  type Unaligned;
  fn to_unaligned(self) -> Self::Unaligned;
  fn from_unaligned(x: Self::Unaligned) -> Self;
}

enum UVecInner<T: AsUnaligned> {
    Owned(Vec<T>),
    Aligned(&'a [T]),
    UnalignedLE(&'a [T::Unaligned]),
}
```

Also, is there an actual need to have `UVecInner::Aligned`? `UnalignedLE` can always accept transmuted `&'a [T]`, right? And our abstraction cannot produce `&[T]` anyway.

As an implementation detail it does not matter if `UnalignedLE` is written as an `&'a [T::Unaligned]` or `&'a [u8]`, but the former is more correct invariants-wise.

This way our abstraction can also subslice to `&[T::Unaligned]` as necessary.

Reply:
Can we do this without `unsafe` code?

My proposed `ByteSliceLE` and `ByteArrayLE` structs are a safe way to convert between aligned and unaligned memory, and they are zero-cost, as shown by the benchmarks and WebAssembly output above.  `T` just needs to implement `Into` and `From` for those types.

`UnalignedLE(&'a [u8])` is roughly equivalent to `UnalignedLE(&'a [ByteSliceLE])`, and I could make that conversion explicit.

I would like to support a way to make specialized versions of `ByteSliceLE` that perform their bit validation during deserialization, which is where https://github.com/zbraniecki/tinystr/issues/41 came from.

Reply:
Before, I was thinking of having two types, `T` and `U`, but a trait with an associated type is a nice way to abstract that.  How about making the trait safe, i.e.,

```rust
trait AsUnaligned {
    type Unaligned;
}
```

and then enforcing the following bounds on the appropriate functions:

```rust
// required for `.get()` and `.iter()`
T: From<T::AsUnaligned::Unaligned>

// required for `.from_unaligned_le_bytes()`
T::AsUnaligned: From<ByteArrayLE<N>>,

// required for `.to_le_bytes()`
T::AsUnaligned: Into<ByteArrayLE<N>>,
```

Reply:
Essentially what we want is

```rust
impl<'a> TryFrom<&'a [[u8; 4]]> for &'a [UU32LE]
```

Reply:
I think we need `unsafe` code if we want stuff to be validity-checked on deserialization; since TinyStr is not a fully POD type with all valid bit patterns. `u32` is.

The unsafe trait is necessary if we want to be allowed to transmute between `&[T]` and `&[U]`. if you want to make it safe, we need the following trait def:

```
unsafe trait AsUnaligned {
  type Unaligned;
  fn to_unaligned(self) -> Self::Unaligned;
  fn from_unaligned(x: Self::Unaligned) -> Self;
  fn to_unaligned_slice(x: &[Self]) -> &[Self::Unaligned];
}
```

And then individual implementors will have to use unsafe code to implement this trait (which they had to anyway). This would make `UVec` fully safe code, too.

I actually prefer this route, come to think of it, it's nice if you don't need to use `unsafe trait`.

Reply:
Where's the conversion from `&[u8]` to `&[Self::Unaligned]` in that model? How do we guarantee safety of that conversion?

Reply:
Why do we need that conversion? For deserialization? I was imagining that that would just happen through a deserialization impl on `&[UU32]` and friends, but if we want to do it generically, we can also just have `trait Unaligned` with `fn from_slice(&[u8]) -> &[Self]`.

Reply:
Done.  Here's what I came up with:

```rust
pub trait ULE
where
    Self: Sized,
{
    type Error;

    fn parse_bytes(bytes: &[u8]) -> Result<&[Self], Self::Error>;
    fn as_bytes(slice: &[Self]) -> &[u8];
}

pub trait AsULE {
    type ULE: ULE;

    fn as_unaligned(&self) -> Self::ULE;
    fn from_unaligned(unaligned: &Self::ULE) -> Self;
}
```

This results in all safe code in UVec.  The only unsafe code is hidden inside the `parse_bytes` and `as_bytes` functions of the ULE trait.